src/processor_cxx17/msg_handler.cxx

```diff
@@ -0,0 +1,126 @@
+#include "msg_handler.hxx"
+
+#include <arpa/inet.h>
+#include <cstring>
+#include <iostream>
+#include <net/if.h>
+#include <optional>
+#include <vector>
+
+#include "base_message.hxx"
+
+namespace cpp11_app {
+
+MessageHandler::MessageHandler(boost::asio::io_context &io_context):
+    net::TcpServerConnection(io_context)
+{
+}
+
+void MessageHandler::cbReceive(const std::vector<std::byte>& data)
+{
+    if (!data.size()) {
+        std::cerr << __func__ << ": buffer is empty" << std::endl;
+        return;
+    }
+
+    // this place should be optimized with PMR
+    std::vector<std::byte> raw_msg(data);
+    std::optional<Message> msg;
+
+    switch (static_cast<msg::MsgType>(getMsgType())) {
+        case msg::MsgType::ControlMsg:
+            msg = msg::ControlMsg<std::vector<std::byte>>(std::move(raw_msg));
+            break;
+        case msg::MsgType::ControlMsgAck:
+            msg = msg::ControlMsgAck<std::vector<std::byte>>(std::move(raw_msg));
+            break;
+        case msg::MsgType::ControlMsgNack:
+            msg = msg::ControlMsgNack<std::vector<std::byte>>(std::move(raw_msg));
+            break;
+        case msg::MsgType::ControlMsgAckCks:
+            msg = msg::ControlMsgAckCks<std::vector<std::byte>>(std::move(raw_msg));
+            break;
+        case msg::MsgType::DataMsg:
+            msg = msg::DataMsg<std::vector<std::byte>>(std::move(raw_msg));
+            break;
+        default:
+            std::cerr << __func__ << ": Wrong msg type: " << std::hex
+                << getMsgType() << std::endl;
+            return;
+    }
+
+    HandleMessage(std::move(msg.value()));
+}
+
+bool MessageHandler::isValidMsgType() const
+{
+    switch (static_cast<msg::MsgType>(getMsgType())) {
+        case msg::MsgType::ControlMsg:
+        case msg::MsgType::ControlMsgAck:
+        case msg::MsgType::ControlMsgNack:
+        case msg::MsgType::ControlMsgAckCks:
+        case msg::MsgType::DataMsg:
+            return true;
+        default:
+            return false;
+    }
+}
+
+uint32_t MessageHandler::getBodySize() const noexcept
+{
+    if (mInputBuffer.size() < (sMsgTypeLen + sMsgSizeLen)) {
+        return 0;
+    }
+
+    uint32_t msg_size = 0;
+
+    std::memcpy(static_cast<void*>(&msg_size),
+            static_cast<const void*>(mInputBuffer.data() + sMsgTypeLen),
+            sizeof(uint32_t));
+
+    return ntohl(msg_size);
+}
+
+uint16_t MessageHandler::getMsgType() const
+{
+    if (mInputBuffer.size() < sMsgTypeLen) {
+        return 0;
+    }
+
+    uint16_t msg_type = 0;
+
+    std::memcpy(static_cast<void*>(&msg_type),
+            static_cast<const void*>(mInputBuffer.data()), sizeof(uint16_t));
+
+    return ntohs(msg_type);
+}
+
+void MessageHandler::HandleMessage(Message msg)
+{
+    std::visit([](auto& arg) {
+                switch (arg.GetMsgType()) {
+                    case msg::MsgType::ControlMsg:
+                        std::cout << "Handling Control Message" << std::endl;
```

Comment:
Let's also add get() for message objects and extract at least 1 byte after the fixed header, to emulate the real processing.

Reply:
done, for CPP11 as well